From ffe03be8d6f2633aa4cba86e874322f14008fec1 Mon Sep 17 00:00:00 2001
From: JW Wang <wangchun@google.com>
Date: Mon, 3 Jan 2022 15:43:54 +0800
Subject: [PATCH] Don't abandon child sessions (1/n)

It will throw if abandon() is called on a child session.

Bug: 211944991
Bug: 67862680
Test: to be added
Change-Id: Ib0ba9f3786dda2d3174f3ea8c65d1061a3fcb586
Merged-In: Ib0ba9f3786dda2d3174f3ea8c65d1061a3fcb586
(cherry picked from commit 8b67e7db79d15ab448ae8f00b40a4a350ab3d537)
(cherry picked from commit c981d90b495955a946bed7113517eeb3d823c88b)
Merged-In:Ib0ba9f3786dda2d3174f3ea8c65d1061a3fcb586
---
 .../java/com/android/server/pm/PackageInstallerService.java | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/services/core/java/com/android/server/pm/PackageInstallerService.java b/services/core/java/com/android/server/pm/PackageInstallerService.java
index 7fdeb27203bd..2f1fa2221e98 100644
--- a/services/core/java/com/android/server/pm/PackageInstallerService.java
+++ b/services/core/java/com/android/server/pm/PackageInstallerService.java
@@ -388,7 +388,11 @@ public class PackageInstallerService extends IPackageInstaller.Stub implements
                 if (age >= MAX_SESSION_AGE_ON_LOW_STORAGE_MILLIS) {
                     // Aggressively close old sessions because we are running low on storage
                     // Their staging dirs will be removed too
-                    session.abandon();
+                    PackageInstallerSession root = !session.hasParentSessionId()
+                            ? session : mSessions.get(session.getParentSessionId());
+                    if (!root.isDestroyed()) {
+                        root.abandon();
+                    }
                 } else {
                     // Session is new enough, so it deserves to be kept even on low storage
                     unclaimedStagingDirsOnVolume.remove(session.stageDir);
-- 
2.33.1

